File: frontend-react/src/utils/structuredLogging.ts
/**
 * Structured Logging Service for the Frontend
 * 
 * This module provides structured logging with OpenTelemetry integration,
 * error tracking, and performance monitoring for the React application.
 */

export interface LogLevel {
  DEBUG: 'debug';
  INFO: 'info';
  WARN: 'warn';
  ERROR: 'error';
}

export interface LogEvent {
  timestamp: string;
  level: keyof LogLevel;
  message: string;
  eventType?: string;
  userId?: string;
  sessionId?: string;
  requestId?: string;
  endpoint?: string;
  method?: string;
  statusCode?: number;
  duration?: number;
  error?: string;
  traceId?: string;
  spanId?: string;
  extra?: Record<string, any>;
}

export interface PerformanceMetric {
  name: string;
  value: number;
  unit?: string;
  tags?: Record<string, string>;
}

export interface ErrorInfo {
  code?: string;
  message: string;
  stack?: string;
  context?: string;
  userAction?: string;
}

class StructuredLogger {
  private sessionId: string;
  private userId?: string;
  private isInitialized = false;
  private logBuffer: LogEvent[] = [];
  private maxBufferSize = 100;
  private flushInterval = 30000; // 30 seconds
  private flushTimer?: NodeJS.Timeout;

  constructor() {
    this.sessionId = this.generateSessionId();
    this.initialize();
  }

  private generateSessionId(): string {
    return `session_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
  }

  private initialize(): void {
    if (this.isInitialized) return;

    // Setup periodic flush
    this.flushTimer = setInterval(() => {
      this.flushLogs();
    }, this.flushInterval);

    // Setup error handlers
    this.setupErrorHandlers();

    // Setup performance monitoring
    this.setupPerformanceMonitoring();

    this.isInitialized = true;
    this.log('info', 'Structured Logger initialized', { eventType: 'logger_init' });
  }

  private setupErrorHandlers(): void {
    // Global error handler
    window.addEventListener('error', (event) => {
      this.log('error', 'Global error caught', {
        eventType: 'global_error',
        error: event.error?.message || event.message,
        filename: event.filename,
        lineno: event.lineno,
        colno: event.colno,
      });
    });

    // Unhandled promise rejection handler
    window.addEventListener('unhandledrejection', (event) => {
      this.log('error', 'Unhandled promise rejection', {
        eventType: 'unhandled_rejection',
        error: event.reason?.message || String(event.reason),
      });
    });
  }

  private setupPerformanceMonitoring(): void {
    // Monitor Core Web Vitals
    if ('PerformanceObserver' in window) {
      // First Contentful Paint
      try {
        const fcpObserver = new PerformanceObserver((list) => {
          const entries = list.getEntries();
          const fcpEntry = entries.find(entry => entry.name === 'first-contentful-paint');
          if (fcpEntry) {
            this.logPerformanceMetric('FCP', fcpEntry.startTime, 'ms');
          }
        });
        fcpObserver.observe({ entryTypes: ['paint'] });
      } catch (e) {
        this.log('warn', 'Failed to setup FCP observer', { error: String(e) });
      }

      // Largest Contentful Paint
      try {
        const lcpObserver = new PerformanceObserver((list) => {
          const entries = list.getEntries();
          const lastEntry = entries[entries.length - 1];
          if (lastEntry) {
            this.logPerformanceMetric('LCP', lastEntry.startTime, 'ms');
          }
        });
        lcpObserver.observe({ entryTypes: ['largest-contentful-paint'] });
      } catch (e) {
        this.log('warn', 'Failed to setup LCP observer', { error: String(e) });
      }

      // First Input Delay
      try {
        const fidObserver = new PerformanceObserver((list) => {
          const entries = list.getEntries();
          entries.forEach((entry: any) => {
            const fid = entry.processingStart - entry.startTime;
            this.logPerformanceMetric('FID', fid, 'ms');
          });
        });
        fidObserver.observe({ entryTypes: ['first-input'] });
      } catch (e) {
        this.log('warn', 'Failed to setup FID observer', { error: String(e) });
      }
    }
  }

  private getTraceContext(): { traceId?: string; spanId?: string } {
    // Try to get trace context from OpenTelemetry if available
    if (typeof window !== 'undefined' && (window as any).otel) {
      const tracer = (window as any).otel.trace.getTracer('frontend');
      const span = tracer.getCurrentSpan();
      if (span) {
        const context = span.getSpanContext();
        return {
          traceId: context.traceId,
          spanId: context.spanId,
        };
      }
    }
    return {};
  }

  private createLogEvent(
    level: keyof LogLevel,
    message: string,
    extra?: Record<string, any>
  ): LogEvent {
    const { traceId, spanId } = this.getTraceContext();
    
    return {
      timestamp: new Date().toISOString(),
      level,
      message,
      sessionId: this.sessionId,
      userId: this.userId,
      traceId,
      spanId,
      ...extra,
    };
  }

  private addToBuffer(event: LogEvent): void {
    this.logBuffer.push(event);
    
    // Flush if buffer is full
    if (this.logBuffer.length >= this.maxBufferSize) {
      this.flushLogs();
    }
  }

  private async flushLogs(): Promise<void> {
    if (this.logBuffer.length === 0) return;

    const logsToFlush = [...this.logBuffer];
    this.logBuffer = [];

    try {
      // Send logs to backend
      await this.sendLogsToBackend(logsToFlush);
    } catch (error) {
      // Fallback to console if backend is unavailable
      console.warn('Failed to send logs to backend, falling back to console:', error);
      logsToFlush.forEach(event => {
        console.log(`[${event.level.toUpperCase()}] ${event.message}`, event);
      });
    }
  }

  private async sendLogsToBackend(logs: LogEvent[]): Promise<void> {
    const response = await fetch('/api/v1/logs/batch', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ logs }),
    });

    if (!response.ok) {
      throw new Error(`Failed to send logs: ${response.status}`);
    }
  }

  public setUserId(userId: string): void {
    this.userId = userId;
    this.log('info', 'User ID set', { eventType: 'user_set', userId });
  }

  public log(
    level: keyof LogLevel,
    message: string,
    extra?: Record<string, any>
  ): void {
    const event = this.createLogEvent(level, message, extra);
    
    // Add to buffer for batch processing
    this.addToBuffer(event);
    
    // Also log to console in development
    if (process.env.NODE_ENV === 'development') {
      const consoleMethod = level === 'error' ? 'error' : 
                           level === 'warn' ? 'warn' : 
                           level === 'info' ? 'info' : 'log';
      console[consoleMethod](`[${level.toUpperCase()}] ${message}`, extra);
    }
  }

  public logApiRequest(
    method: string,
    endpoint: string,
    statusCode: number,
    duration: number,
    requestId?: string,
    requestSize?: number,
    responseSize?: number,
    error?: string
  ): void {
    this.log('info', `API Request: ${method} ${endpoint}`, {
      eventType: 'api_request',
      method,
      endpoint,
      statusCode,
      duration,
      requestId,
      requestSize,
      responseSize,
      error,
    });
  }

  public logDatabaseQuery(
    queryType: string,
    tableName: string,
    duration: number,
    rowsAffected?: number,
    error?: string
  ): void {
    this.log('debug', `Database Query: ${queryType} on ${tableName}`, {
      eventType: 'database_query',
      queryType,
      tableName,
      duration,
      rowsAffected,
      error,
    });
  }

  public logSecurityEvent(
    eventType: string,
    description: string,
    userId?: string,
    ipAddress?: string,
    severity: keyof LogLevel = 'info',
    details?: Record<string, any>
  ): void {
    this.log(severity, `Security Event: ${description}`, {
      eventType: 'security_event',
      securityEventType: eventType,
      userId,
      ipAddress,
      details,
    });
  }

  public logPerformanceMetric(
    name: string,
    value: number,
    unit?: string,
    tags?: Record<string, string>
  ): void {
    this.log('info', `Performance Metric: ${name}`, {
      eventType: 'performance_metric',
      metricName: name,
      value,
      unit,
      tags,
    });
  }

  public logError(
    error: Error | string,
    context?: string,
    userAction?: string
  ): void {
    const errorInfo: ErrorInfo = {
      message: typeof error === 'string' ? error : error.message,
      stack: typeof error === 'string' ? undefined : error.stack,
      context,
      userAction,
    };

    this.log('error', `Error: ${errorInfo.message}`, {
      eventType: 'error',
      error: errorInfo,
    });
  }

  public async traceOperation<T>(
    operationName: string,
    operation: () => Promise<T>,
    attributes?: Record<string, any>
  ): Promise<T> {
    const startTime = performance.now();
    
    try {
      this.log('debug', `Operation started: ${operationName}`, {
        eventType: 'operation_start',
        operationName,
        ...attributes,
      });

      const result = await operation();
      
      const duration = performance.now() - startTime;
      this.log('debug', `Operation completed: ${operationName}`, {
        eventType: 'operation_success',
        operationName,
        duration,
        ...attributes,
      });

      return result;
    } catch (error) {
      const duration = performance.now() - startTime;
      this.log('error', `Operation failed: ${operationName}`, {
        eventType: 'operation_error',
        operationName,
        duration,
        error: error instanceof Error ? error.message : String(error),
        ...attributes,
      });
      throw error;
    }
  }

  public destroy(): void {
    if (this.flushTimer) {
      clearInterval(this.flushTimer);
    }
    this.flushLogs();
  }
}

// Global logger instance
export const structuredLogger = new StructuredLogger();

// Convenience functions
export const log = (level: keyof LogLevel, message: string, extra?: Record<string, any>) => {
  structuredLogger.log(level, message, extra);
};

export const logApiRequest = (
  method: string,
  endpoint: string,
  statusCode: number,
  duration: number,
  requestId?: string,
  requestSize?: number,
  responseSize?: number,
  error?: string
) => {
  structuredLogger.logApiRequest(method, endpoint, statusCode, duration, requestId, requestSize, responseSize, error);
};

export const logError = (error: Error | string, context?: string, userAction?: string) => {
  structuredLogger.logError(error, context, userAction);
};

export const logPerformanceMetric = (name: string, value: number, unit?: string, tags?: Record<string, string>) => {
  structuredLogger.logPerformanceMetric(name, value, unit, tags);
};

export const traceOperation = <T>(
  operationName: string,
  operation: () => Promise<T>,
  attributes?: Record<string, any>
) => {
  return structuredLogger.traceOperation(operationName, operation, attributes);
};

export const setUserId = (userId: string) => {
  structuredLogger.setUserId(userId);
};

// Cleanup on page unload
if (typeof window !== 'undefined') {
  window.addEventListener('beforeunload', () => {
    structuredLogger.destroy();
  });
}

File: backend/app/core/structured_logging.py
"""
Structured Logging Service for the AI Assistant Platform.

This module provides structured logging with OpenTelemetry integration,
log correlation, and standardized log formats.
"""

import json
import logging
import sys
from contextlib import contextmanager
from datetime import UTC, datetime
from typing import Any

from loguru import logger
from opentelemetry import trace
from opentelemetry.trace import Status, StatusCode

from backend.app.core.config import get_settings


class StructuredLogger:
    """Structured logger with OpenTelemetry integration."""

    def __init__(self):
        self.settings = get_settings()
        self._setup_loguru()
        self._setup_standard_logging()

    def _setup_loguru(self):
        """Setup Loguru with structured logging."""
        # Remove default handler
        logger.remove()

        # Add structured console handler
        logger.add(
            sys.stdout,
            format=self._format_log,
            level=self.settings.log_level,
            serialize=True,
            backtrace=True,
            diagnose=True,
        )

        # Add file handler if configured
        if self.settings.log_file and self.settings.log_file != "./logs/app.log":
            logger.add(
                self.settings.log_file,
                format=self._format_log,
                level=self.settings.log_level,
                serialize=True,
                rotation="10 MB",
                retention="30 days",
                compression="gz",
            )

    def _setup_standard_logging(self):
        """Setup standard Python logging to work with Loguru."""
        logging.basicConfig(handlers=[InterceptHandler()], level=0, force=True)

    def _format_log(self, record):
        """Format log record with structured data."""
        # Get current span for correlation
        current_span = trace.get_current_span()
        span_context = current_span.get_span_context() if current_span else None

        # Handle both Loguru and standard Python logging records
        if isinstance(record, dict):
            # Loguru record format (dictionary)
            if "time" in record and hasattr(record["time"], "timestamp"):
                timestamp = datetime.fromtimestamp(record["time"].timestamp(), tz=UTC).isoformat()
            else:
                timestamp = datetime.now(UTC).isoformat()
            level = record.get("level", {}).name if isinstance(record.get("level"), object) else str(record.get("level", "INFO"))
            logger_name = record.get("name", "unknown")
            message = record.get("message", "")
            module = record.get("module", "unknown")
            function = record.get("function", "unknown")
            line = record.get("line", 0)
        elif hasattr(record, "time") and hasattr(record.time, "timestamp"):
            # Loguru record format (object)
            timestamp = datetime.fromtimestamp(record.time.timestamp(), tz=UTC).isoformat()
            level = record.level.name
            logger_name = record.name
            message = record.message
            module = record.module
            function = record.function
            line = record.line
        else:
            # Standard Python logging record format
            timestamp = datetime.fromtimestamp(record.created, tz=UTC).isoformat()
            level = record.levelname
            logger_name = record.name
            message = record.getMessage()
            module = record.module
            function = record.funcName
            line = record.lineno

        # Base log structure
        log_data = {
            "timestamp": timestamp,
            "level": level,
            "logger": logger_name,
            "message": message,
            "module": module,
            "function": function,
            "line": line,
        }

        # Add trace correlation if available
        if span_context and span_context.is_valid:
            log_data.update(
                {
                    "trace_id": format(span_context.trace_id, "032x"),
                    "span_id": format(span_context.span_id, "016x"),
                }
            )

        # Add extra fields from record
        if hasattr(record, "extra") and record.extra:
            log_data.update(record.extra)

        # Add exception info if present
        if hasattr(record, "exception") and record.exception:
            log_data["exception"] = {
                "type": record.exception.type.__name__,
                "value": str(record.exception.value),
                "traceback": record.exception.traceback,
            }

        return json.dumps(log_data, ensure_ascii=False, default=str)

    def log_event(
        self,
        level: str,
        message: str,
        event_type: str = None,
        user_id: str = None,
        session_id: str = None,
        request_id: str = None,
        endpoint: str = None,
        method: str = None,
        status_code: int = None,
        duration: float = None,
        extra: dict[str, Any] = None,
        **kwargs,
    ):
        """Log a structured event."""
        log_data = {
            "event_type": event_type,
            "user_id": user_id,
            "session_id": session_id,
            "request_id": request_id,
            "endpoint": endpoint,
            "method": method,
            "status_code": status_code,
            "duration": duration,
        }

        # Add extra data
        if extra:
            log_data.update(extra)

        # Add additional kwargs
        log_data.update(kwargs)

        # Remove None values
        log_data = {k: v for k, v in log_data.items() if v is not None}

        # Log with structured data
        getattr(logger, level.lower())(message, extra=log_data)

    def log_api_request(
        self,
        method: str,
        endpoint: str,
        status_code: int,
        duration: float,
        user_id: str = None,
        request_id: str = None,
        request_size: int = None,
        response_size: int = None,
        error: str = None,
    ):
        """Log API request with structured data."""
        self.log_event(
            level="INFO",
            message=f"API Request: {method} {endpoint}",
            event_type="api_request",
            method=method,
            endpoint=endpoint,
            status_code=status_code,
            duration=duration,
            user_id=user_id,
            request_id=request_id,
            request_size=request_size,
            response_size=response_size,
            error=error,
        )

    def log_database_query(
        self,
        query_type: str,
        table_name: str,
        duration: float,
        rows_affected: int = None,
        error: str = None,
    ):
        """Log database query with structured data."""
        self.log_event(
            level="DEBUG",
            message=f"Database Query: {query_type} on {table_name}",
            event_type="database_query",
            query_type=query_type,
            table_name=table_name,
            duration=duration,
            rows_affected=rows_affected,
            error=error,
        )

    def log_security_event(
        self,
        event_type: str,
        description: str,
        user_id: str = None,
        ip_address: str = None,
        severity: str = "INFO",
        details: dict[str, Any] = None,
    ):
        """Log security event with structured data."""
        self.log_event(
            level=severity.upper(),
            message=f"Security Event: {description}",
            event_type="security_event",
            security_event_type=event_type,
            user_id=user_id,
            ip_address=ip_address,
            details=details,
        )

    def log_performance_metric(
        self,
        metric_name: str,
        value: float,
        unit: str = None,
        tags: dict[str, str] = None,
    ):
        """Log performance metric with structured data."""
        self.log_event(
            level="INFO",
            message=f"Performance Metric: {metric_name}",
            event_type="performance_metric",
            metric_name=metric_name,
            value=value,
            unit=unit,
            tags=tags,
        )

    @contextmanager
    def trace_operation(
        self,
        operation_name: str,
        attributes: dict[str, Any] = None,
        user_id: str = None,
        request_id: str = None,
    ):
        """Context manager for tracing operations with logging."""
        tracer = trace.get_tracer(__name__)

        with tracer.start_as_current_span(
            operation_name, attributes=attributes
        ) as span:
            try:
                # Log operation start
                self.log_event(
                    level="DEBUG",
                    message=f"Operation started: {operation_name}",
                    event_type="operation_start",
                    operation_name=operation_name,
                    user_id=user_id,
                    request_id=request_id,
                )

                yield span

                # Log operation success
                span.set_status(Status(StatusCode.OK))
                self.log_event(
                    level="DEBUG",
                    message=f"Operation completed: {operation_name}",
                    event_type="operation_success",
                    operation_name=operation_name,
                    user_id=user_id,
                    request_id=request_id,
                )

            except Exception as e:
                # Log operation error
                span.set_status(Status(StatusCode.ERROR, str(e)))
                span.record_exception(e)
                self.log_event(
                    level="ERROR",
                    message=f"Operation failed: {operation_name}",
                    event_type="operation_error",
                    operation_name=operation_name,
                    user_id=user_id,
                    request_id=request_id,
                    error=str(e),
                )
                raise


class InterceptHandler(logging.Handler):
    """Intercept standard logging and redirect to Loguru."""

    def emit(self, record):
        try:
            level = logger.level(record.levelname).name
        except ValueError:
            level = record.levelno

        frame, depth = logging.currentframe(), 2
        while frame.f_code.co_filename == logging.__file__:
            frame = frame.f_back
            depth += 1

        logger.opt(depth=depth, exception=record.exc_info).log(
            level, record.getMessage()
        )


# Global structured logger instance (disabled to avoid conflicts)
# structured_logger = StructuredLogger()
structured_logger = None


def get_logger(name: str = None) -> StructuredLogger:
    """Get structured logger instance."""
    return structured_logger


def log_event(
    level: str,
    message: str,
    event_type: str = None,
    user_id: str = None,
    session_id: str = None,
    request_id: str = None,
    endpoint: str = None,
    method: str = None,
    status_code: int = None,
    duration: float = None,
    extra: dict[str, Any] = None,
    **kwargs,
):
    """Log a structured event."""
    if structured_logger is not None:
        structured_logger.log_event(
            level=level,
            message=message,
            event_type=event_type,
            user_id=user_id,
            session_id=session_id,
            request_id=request_id,
            endpoint=endpoint,
            method=method,
            status_code=status_code,
            duration=duration,
            extra=extra,
            **kwargs,
        )
    else:
        # Fallback to basic logging
        logger.log(level.upper(), message)


def log_api_request(
    method: str,
    endpoint: str,
    status_code: int,
    duration: float,
    user_id: str = None,
    request_id: str = None,
    request_size: int = None,
    response_size: int = None,
    error: str = None,
):
    """Log API request with structured data."""
    structured_logger.log_api_request(
        method=method,
        endpoint=endpoint,
        status_code=status_code,
        duration=duration,
        user_id=user_id,
        request_id=request_id,
        request_size=request_size,
        response_size=response_size,
        error=error,
    )


def log_database_query(
    query_type: str,
    table_name: str,
    duration: float,
    rows_affected: int = None,
    error: str = None,
):
    """Log database query with structured data."""
    structured_logger.log_database_query(
        query_type=query_type,
        table_name=table_name,
        duration=duration,
        rows_affected=rows_affected,
        error=error,
    )


def log_security_event(
    event_type: str,
    description: str,
    user_id: str = None,
    ip_address: str = None,
    severity: str = "INFO",
    details: dict[str, Any] = None,
):
    """Log security event with structured data."""
    structured_logger.log_security_event(
        event_type=event_type,
        description=description,
        user_id=user_id,
        ip_address=ip_address,
        severity=severity,
        details=details,
    )


def log_performance_metric(
    metric_name: str,
    value: float,
    unit: str = None,
    tags: dict[str, str] = None,
):
    """Log performance metric with structured data."""
    structured_logger.log_performance_metric(
        metric_name=metric_name,
        value=value,
        unit=unit,
        tags=tags,
    )


def trace_operation(
    operation_name: str,
    attributes: dict[str, Any] = None,
    user_id: str = None,
    request_id: str = None,
):
    """Context manager for tracing operations with logging."""
    return structured_logger.trace_operation(
        operation_name=operation_name,
        attributes=attributes,
        user_id=user_id,
        request_id=request_id,
    )
